control.lua

```diff
@@ -33,24 +35,118 @@ script.on_configuration_changed(function(data)
+  local name = entity.backer_name
+  local id = entity.unit_number
+  if entity.name == "train-stop-dispatcher" or name:match("%.[123456789]%d*$") then
     if not global.stations[name] then
       global.stations[name] = {}
+      global.stations[name][id] = entity
+      debug("Added first station named: ", name)
+    else
+      global.stations[name][id] = entity
+      debug("Added station: ", name)
     end
-    table.insert(global.stations[name], event.created_entity)
-    debug("New station: ", name)
+  else
+    debug("Ignoring new station: ", name)
   end
 end
-script.on_event(defines.events.on_built_entity, entity_built)
-script.on_event(defines.events.on_robot_built_entity, entity_built)
 
+-- Remove station from global.stations if it is in the list
+function remove_station(entity, old_name)
+  local name = old_name
+  if not name then name = entity.backer_name end
+  local id = entity.unit_number
+  if global.stations[name] then
+    if global.stations[name][id] then
+      global.stations[name][id] = nil
+      if table_size(global.stations[name]) == 0 then
+        global.stations[name] = nil
+        debug("Removed last station named: ", name)
+      else
+        debug("Removed station: ", name)
+      end
+    end
+  end
+end
+
+-- Add stations when built/revived
+function entity_built(event)
+  local entity = event.created_entity
+  if not entity then entity = event.entity end
+  if entity and entity.valid then
+    if entity.type == "train-stop" then
+      add_station(entity)
+    end
+  end
+end
+script.on_event({ defines.events.on_built_entity,
+                  defines.events.on_robot_built_entity,
+                  defines.events.script_raised_built,
+                  defines.events.script_raised_revive },
+                entity_built)
+
+-- Remove station when mined/destroyed
+function entity_removed(event)
+  local entity = event.entity
+  if entity and entity.valid then
+    if entity.type == "train-stop" then
+      remove_station(entity)
+    end
+  end
+end
+script.on_event({ defines.events.on_player_mined_entity,
+                  defines.events.on_robot_mined_entity,
+                  defines.events.on_entity_died,
+                  defines.events.script_raised_destroy },
+                entity_removed)
+
+-- Update station when renamed by player or script
+function entity_renamed(event)
+  local entity = event.entity
+  if entity and entity.valid then
+    if entity.type == "train-stop" then
+      remove_station(entity, event.old_name)
+      add_station(entity)
+    end
+  end
+end
+script.on_event(defines.events.on_entity_renamed, entity_renamed)
+
+
+-- Build list of stations
+function build_list_stations()
+  global.stations = {}
+  local stations = game.surfaces["nauvis"].find_entities_filtered{type= "train-stop"}
+  for _, station in pairs(stations) do
+    add_station(station)
+  end
+  debug("Stations list rebuilt")
+end
+
+
+-- Scrub list of trains
+function scrub_trains()
```

Comment:
I put it in the migration code because the old code sometimes left invalid trains in the queues and this will remove them during the upgrade process.  I don't know if there are any edge cases in the new code that will produce orphaned train entries also (for example, loading the map after uninstalling some modded trains, and an entire Awaiting Dispatch train disappears).

Reply:
That is what I hate the most with Factorio modding, having to maintain these tables of objects that may be changed by the game or mods in tons of way. Let's keep it that way for now.